### PR TITLE
[Performance] Improve SpecifyIFormatProvider (CA1305) performance

### PR DESCRIPTION
**Pull request by Youssef1313**

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs [hunk 1]
```diff
@@ -266,19 +266,16 @@ protected override void InitializeWorker(CompilationStartAnalysisContext context
...
                     {
-                        var argument = invocationExpression.Arguments[index];
-
-                        if (argument != null && currentUICultureProperty != null &&
-                            installedUICultureProperty != null && currentThreadCurrentUICultureProperty != null)
+                        if (!iformatProviderType.Equals(argument.Parameter?.Type))
                         {
-                            var semanticModel = argument.SemanticModel!;
-
-                            var symbol = semanticModel.GetSymbolInfo(argument.Value.Syntax, oaContext.CancellationToken).Symbol;
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs:279]: @mavasani I assume semantic model calls are more expensive than getting info directly from IOperation?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs [hunk 2]
```diff
@@ -186,11 +186,11 @@ protected override void InitializeWorker(CompilationStartAnalysisContext context
 
                 #region "IFormatProviderAlternateStringRule Only"
                 if (stringFormatMemberWithIFormatProviderStringAndParamsObjectParameter != null &&
-                    !oaContext.Options.IsConfiguredToSkipAnalysis(IFormatProviderAlternateStringRule, targetMethod, oaContext.ContainingSymbol, oaContext.Compilation) &&
                     (targetMethod.Equals(stringFormatMemberWithStringAndObjectParameter) ||
                      targetMethod.Equals(stringFormatMemberWithStringObjectAndObjectParameter) ||
                      targetMethod.Equals(stringFormatMemberWithStringObjectObjectAndObjectParameter) ||
-                     targetMethod.Equals(stringFormatMemberWithStringAndParamsObjectParameter)))
+                     targetMethod.Equals(stringFormatMemberWithStringAndParamsObjectParameter)) &&
+                    !oaContext.Options.IsConfiguredToSkipAnalysis(IFormatProviderAlternateStringRule, targetMethod, oaContext.ContainingSymbol, oaContext.Compilation))
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs:193]: Symbol equality should be faster than `IsConfiguredToSkipAnalysis`

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs [hunk 3]
```diff
@@ -324,14 +321,6 @@ static IEnumerable<IMethodSymbol> GetToStringMethods(INamedTypeSymbol namedTypeS
                 => namedTypeSymbol.GetMembers("ToString").OfType<IMethodSymbol>().WhereNotNull();
         }
 
-        private static IEnumerable<int> GetIndexesOfParameterType(IMethodSymbol targetMethod, INamedTypeSymbol formatProviderType)
-        {
-            return targetMethod.Parameters
-                .Select((Parameter, Index) => (Parameter, Index))
-                .Where(x => x.Parameter.Type.Equals(formatProviderType))
-                .Select(x => x.Index);
-        }
```
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs:333]: I'm now looping over arguments directly and than checking if the parameter is of interest. The new implementation should hopefully be more performant and more straightforward.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs (hunk 1 above)
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyIFormatProvider.cs:279]: Yes, definitely.

**Review by mavasani — APPROVED** [review]

**Comment by mavasani** [comment]
@buyaa-n for another review

**Review by buyaa-n — APPROVED** [review]
Looks good, thanks!

**Pull request merged**